gap/oper.gi

```diff
@@ -1725,6 +1725,146 @@ function(D, list)
   return DigraphShortestDistance(D, list[1], list[2]);
 end);
 
+InstallMethod(DominatorTree, "for a digraph and a vertex",
+[IsDigraph, IsPosInt],
+function(D, root)
+  local M, node_to_preorder_num, preorder_num_to_node, parent, index, next,
+  current, succ, prev, n, semi, lastlinked, label, bucket, idom,
+  compress, eval, pred, N, w, y, x, i, v;
+
+  M := DigraphNrVertices(D);
+
+  node_to_preorder_num := [];
+  node_to_preorder_num[root] := 1;
+  preorder_num_to_node := [root];
+
+  parent := [];
+  parent[root] := fail;
+
+  index := ListWithIdenticalEntries(M, 1);
+
+  next := 2;
+  current := root;
+  succ := OutNeighbours(D);
+
+  # Step 1: DFS to establish preorder
+  repeat
+    prev := current;
+    for i in [index[current] .. Length(succ[current])] do
+      n := succ[current][i];
+      if not IsBound(node_to_preorder_num[n]) then
+        Add(preorder_num_to_node, n);
+        parent[n] := current;
+        index[current] := i + 1;
+        node_to_preorder_num[n] := next;
+        next := next + 1;
+        current := n;
+        break;
+      fi;
+    od;
+    # continues from here
+    if prev = current then
+      # we backtrack
+      current := parent[current];
+    fi;
+  until current = fail;
+
+  # Step 2: find semidominators, and first pass of immediate dominators
+  semi := [1 .. M];
+  lastlinked := M + 1;  # never linked
+  label := [];
+  bucket := List([1 .. M], x -> []);
+  idom := [];
+  idom[root] := root;
+
+  compress := function(v)
```

Comment:
This function is recursive and I remember that we said that functions in a package shouldn't be recursive, since there is a limit on how many recursions GAP can "handle".

Reply:
See https://github.com/digraphs/Digraphs/pull/336#issuecomment-755413562